src/protocol/oprf/mod.rs

```diff
@@ -0,0 +1,529 @@
+    for (record_id, row) in items.into_iter().enumerate() {
+        send_channel.send(RecordId::from(record_id), row).await?;
+    }
+    Ok(())
+}
+
+async fn receive_from_peer<C: Context>(
+    ctx: &C,
+    step: &OPRFShuffleStep,
+    direction: Direction,
+    batch_size: u32,
+) -> Result<Vec<OPRFShuffleSingleShare>, Error> {
+    let role = ctx.role().peer(direction);
+    let receive_channel: ReceivingEnd<OPRFShuffleSingleShare> = ctx.narrow(step).recv_channel(role);
+
+    let mut output: Vec<OPRFShuffleSingleShare> = Vec::with_capacity(batch_size as usize);
+    for record_id in 0..batch_size {
+        let msg = receive_channel.receive(RecordId::from(record_id)).await?;
+        output.push(msg);
+    }
+
+    Ok(output)
+}
+
+async fn exchange_c_hat<C: Context, I: IntoIterator<Item = OPRFShuffleSingleShare>>(
+    ctx: &C,
+    batch_size: u32,
+    part_to_send: I,
+) -> Result<Vec<OPRFShuffleSingleShare>, Error> {
+    let (step_send, step_recv, dir) = match ctx.role() {
+        Role::H2 => (
+            OPRFShuffleStep::TransferCHat1,
+            OPRFShuffleStep::TransferCHat2,
+            Direction::Right,
+        ),
+        Role::H3 => (
+            OPRFShuffleStep::TransferCHat2,
+            OPRFShuffleStep::TransferCHat1,
+            Direction::Left,
+        ),
+        role @ Role::H1 => {
+            unreachable!("Role {:?} does not participate in C_hat computation", role)
+        }
+    };
+
+    let ((), received_part) = try_join!(
+        send_to_peer(ctx, &step_send, dir, part_to_send),
+        receive_from_peer(ctx, &step_recv, dir, batch_size),
+    )?;
+
+    Ok(received_part)
+}
+
+// --------------------------- permutation-related function --------------------------------------------- //
+
+fn generate_permutations_with_peers<C: Context>(batch_size: u32, ctx: &C) -> (Vec<u32>, Vec<u32>) {
+    let (step_left, step_right) = match &ctx.role() {
+        Role::H1 => (OPRFShuffleStep::GeneratePi31, OPRFShuffleStep::GeneratePi12),
+        Role::H2 => (OPRFShuffleStep::GeneratePi12, OPRFShuffleStep::GeneratePi23),
+        Role::H3 => (OPRFShuffleStep::GeneratePi23, OPRFShuffleStep::GeneratePi12),
+    };
+
+    let with_left = generate_pseudorandom_permutation(batch_size, ctx, &step_left, Direction::Left);
+    let with_right =
+        generate_pseudorandom_permutation(batch_size, ctx, &step_right, Direction::Right);
+    (with_left, with_right)
+}
+
+fn generate_pseudorandom_permutation<C: Context>(
+    batch_size: u32,
+    ctx: &C,
+    step: &OPRFShuffleStep,
+    with_peer_on_the: Direction,
+) -> Vec<u32> {
+    let narrow_context = ctx.narrow(step);
+    let rng = narrow_context.prss_rng();
+    let mut rng = match with_peer_on_the {
+        Direction::Left => rng.0,
+        Direction::Right => rng.1,
+    };
+
+    let mut permutation = (0..batch_size).collect::<Vec<_>>();
+    permutation.shuffle(&mut rng);
+    permutation
+}
+
+fn permute(
```

Comment:
this relies on the `IntoInterator` specialization to avoid extra allocations, but I don't see why it can't be a one liner `apply(permutation, &mut data)` on the caller's side